---
title: "EigenGWAS Report"
author: "`r Sys.time()`"
#date: "`r Sys.time()`"
output:
 html_document:
    theme: united
    highlight: tango
    code_folding: hide
params:
  froot: arab
  uploadfile: arab
  proportion: 0.2
  espace: 2
  sc: 1
  pcut: 0.05
  height: 480
  width: 480
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.align = 'center')
#source("helper.R")
```

```{r, manhattan,echo=FALSE}
manhattan <- function(dataframe, colors=c("gray10", "gray50"), ymax="max", limitchromosomes=NULL, suggestiveline=-log10(1e-5), genomewideline=NULL, title="", annotate=NULL, ...) {
  
  d=dataframe
  if (!("CHR" %in% names(d) & "BP" %in% names(d) & "P" %in% names(d))) stop("Make sure your data frame contains columns CHR, BP, and P")
  if (!is.null(limitchromosomes)) {
    d=d[d$CHR %in% limitchromosomes, ]
  }
  
  d=subset(na.omit(d[order(d$CHR, d$BP), ]), (P>0 & P<=1)) # remove na's, sort, and keep only 0<P<=1
  d$logp = -log10(d$P)
  d$pos=NA
  ticks=NULL
  lastbase=0 #  colors <- rep(colors,max(d$CHR))[1:max(d$CHR)]
  colors <- rep(colors,max(d$CHR))[1:length(unique(d$CHR))]
  
  if (ymax=="max") ymax<-ceiling(max(d$logp))
  if (ymax<8) ymax<-8
  numchroms=length(unique(d$CHR))
  if (numchroms==1) {
    d$pos=d$BP
    ticks=floor(length(d$pos))/2+1
  } else {
    Uchr=unique(d$CHR)
    for (i in 1:length(Uchr)) {
      if (i==1) {
        d[d$CHR==Uchr[i], ]$pos=d[d$CHR==Uchr[i], ]$BP
      } else {
        lastbase=lastbase+tail(subset(d, CHR==Uchr[i-1])$BP, 1)
        d[d$CHR==Uchr[i], ]$pos=d[d$CHR==Uchr[i], ]$BP+lastbase
      }
      ticks=c(ticks, d[d$CHR==Uchr[i], ]$pos[floor(length(d[d$CHR==Uchr[i], ]$pos)/2)+1])
    }
  }
  if (numchroms==1) {
    with(d, plot(main=title, pos, logp, ylim=c(0,ymax), ylab=expression(-log[10](italic(p))), xlab=paste("Chromosome",unique(d$CHR),"position"), ...))
  } else {
    with(d, plot(main=title, pos, logp, ylim=c(0,ymax), ylab=expression(-log[10](italic(p))), xlab="Chromosome", xaxt="n", type="n", ...))
    axis(1, at=ticks, lab=unique(d$CHR), ...)
    icol=1
    Uchr=unique(d$CHR)
    for (i in 1:length(Uchr)) {
      with(d[d$CHR==Uchr[i], ], points(pos, logp, col=colors[icol], ...))
      icol=icol+1
    }
  }
  if (!is.null(annotate)) {
    d.annotate=d[which(d$SNP %in% annotate), ]
    with(d.annotate, points(pos, logp, col="green3", ...))
  }
  #  if (suggestiveline) abline(h=suggestiveline, col="blue")
  if (!is.null(genomewideline)) {
    abline(h=genomewideline, col="gray")
  } else {
    abline(h=-log10(0.05/nrow(d)), col="gray")    
  }
}
```

# {.tabset .tabset-fade .tabset-pills}
## 1 MAF
```{r, maf,echo=TRUE}
fq=read.table(paste0(params$froot, ".frq"), as.is = T, header = T)
hist(fq$MAF, main="MAF", xlab="Minor allele frequency", xlim=c(0, 0.5), breaks = 50)
```
Minor allele frequency distribution of the sample.

## 2 Eigenanalysis
```{r, pca,echo=TRUE}
layout(matrix(1:2, 1, 2))
pcRun = 10
evalF=read.table(paste0(params$froot, ".eigenval"), as.is = T)
evalF=evalF[,1]
names(evalF) = c(1:pcRun)
barplot(evalF/params$sc, border = F, main="Eigenvalue",ylim = c(0,max(evalF/params$sc)*1.2),xlab = 'Eigenspace')
abline(h=1, lty=2, col="black")

pcF=read.table(paste0(params$froot, ".eigenvec"), as.is = T)
plot(main="eSpace 1 vs 2", pcF[,3], pcF[,4], xlab="eSpace 1", ylab="eSpace 2", bty='n', pch=16, cex=0.5, col=ifelse(pcF[,3]<0, "red", "blue"))
```
The distribution of the top-ranked eigenvalues and the projection of the samples in the first two eigenspaces.

## 3 GRM
```{r, grm,echo=TRUE}
layout(matrix(1:2, 1, 2))
gz=gzfile(paste0(params$froot, ".grm.gz"))
grm=read.table(gz, as.is = T)
offDiag = grm[grm[,1]!=grm[,2], 4]
Ne=-1/mean(offDiag/params$sc)
Me=1/var(offDiag/params$sc)
hist(offDiag/params$sc, main="Pairwise relatedness ", xlab="Self-Relatedness", breaks = 50)
        
nn=nrow(read.table(paste0(params$froot, ".fam"), as.is = T))
mm=nrow(read.table(paste0(params$froot, ".bim"), as.is = T))
legend("topright", legend = c(paste0("ne=", format(Ne, digits=3, nsmall=2), ' [',nn, ']'), paste0("me=", format(Me, digits=3, nsmall=2), ' [',mm,']')), bty='n')

hist(grm[grm[,1]==grm[,2],4]/params$sc, main="Diagonal relatedness", xlab="Relatedness score", breaks = 15)
```
Relatedness score is defined as the pairwise relatedness for any pair of individuals as measured over genome-wide markers. It is often employed for the estimation for additive genetic variance, see VanRaden ([$J Dairy Sci, 2008, 91:4414-4423$](https://www.sciencedirect.com/science/article/pii/S0022030208709901)) for more details. $n_e$ is the effective sample size. If the samples are related to each other much, $n_e$ would be smaller than the real sample size (in square brackets). $m_e$ is the effective number of markers. When markers are in linkage equilibrium, $m_e$=$m$, the number of markers in study (in square brackets). Of note, when the sample has experienced, recent, strong selection, me can be very small, say less than 0.01$m$; however, it can be of demographic factors possible.

## 4 Eigenvalue vs $\lambda_{GC}$
```{r, ev,echo=TRUE}
PC=params$espace
Evev=read.table(paste0(params$froot, ".eigenval"), as.is = T)
GC=array(0, dim=PC)
for(i in 1:PC) {
  eg = read.table(paste0(params$froot, ".", i, ".assoc.linear"), as.is = T, header = T)
  GC[i] = qchisq(median(eg$P, na.rm = T), 1, lower.tail = F)/qchisq(0.5, 1)
}

egc=matrix(c(Evev[1:PC,1]/params$sc, GC), PC, 2, byrow = F)
egc_table = cbind(as.character(seq(1, PC)),as.data.frame(format(egc,digits = 3)))
library(knitr)
kable(egc_table, align = 'c',col.names = c("Eigenspace","Eigenvalue", "Genomic inflation factor"))
rownames(egc)=seq(1, PC)
barplot(t(egc), beside = T, border = F, xlab="eSpace", ylim=c(0,max(egc)+2))
        abline(h=1, lty=2, lwd=2)
legend("topright", legend = c("Eigenvalue", expression(lambda[gc])), pch=15, col=c("black", "grey"), bty='n')

```
Eigenvalue follows a mixture distribution $\eta$$F_{st,s}$+(1-$\eta$)$F_{st,d}$, and $\lambda_{GC}$ is proportional to $F_{st,d}$. If eigenvalue is far larger than $\lambda_{GC}$, it indicates the presence of selection sweep of the sample.

## 5 EigenGWAS
Manhattan plot and Q-Q plot for EigenGWAS scanning on the chosen eigenvector. The p-value threshold, grey line in the Manhattan plot, is set at genome-wide control for type I error rate of 0.05 (Bonferroni correction). The grey points in the Q-Q plot are for the test statistics without technical correction for genome-wide drift, and the dark ones with technical correction for genome-wide drift.
```{r, eGWAS, results="asis",fig.align="center",echo=TRUE}
PC=params$espace
froot = params$froot
tophit = data.frame()

for(i in 1:PC) {
  EigenRes=read.table(paste0(froot, ".", i, ".assoc.linear"), as.is = T, header = T)
  EigenRes=EigenRes[which(!is.na(EigenRes$P)),]
  EigenRes$Praw=EigenRes$P
  gc=qchisq(median(EigenRes$P), 1, lower.tail = F)/qchisq(0.5, 1, lower.tail = F)
  EigenRes$P=pchisq((EigenRes$STAT)^2/gc, 1, lower.tail = F)
  
  EigenResName = paste0(tempdir(),'/EgE',i,'.png')
  QQPlotName = paste0(tempdir(),'/QQE',i,'.png')
  if(!file.exists(EigenResName)){
    EigenRes_sig = EigenRes[EigenRes$P<=0.05,]
    EigenRes_no_sig = EigenRes[EigenRes$P>0.05,]
    EigenRes_com = rbind(EigenRes_sig,EigenRes_no_sig[sample(1:nrow(EigenRes_no_sig),0.6*nrow(EigenRes_no_sig)),])
    chiseq=qchisq(seq(1/nrow(EigenRes_com), 1-1/nrow(EigenRes_com), length.out = nrow(EigenRes_com)), 1)
    
    png(filename = EigenResName,width = params$width,height = params$height)
    manhattan(EigenRes_com, genomewideline = -log10(params$pcut/nrow(EigenRes)), title=paste("eSpace ", i), pch=16, cex=0.3, bty='n')
    dev.off()
    
    png(filename = QQPlotName,width = params$width,height = params$height)
    qqplot(chiseq, qchisq(EigenRes_com$Praw, 1, lower.tail = F), xlab=expression(paste("Theoretical ", chi[1]^2)), ylab=expression(paste("Observed ", chi[1]^2)), bty="n", col="grey", pch=16, cex=0.5)
    points(sort(chiseq), sort(qchisq(EigenRes_com$P, 1, lower.tail = F)), col="black", pch=16, cex=0.5)
    legend("topleft", legend = c("Raw", expression(lambda[gc])), pch=16, cex=1, col=c("grey", "black"), bty='n')
    abline(a=0, b=1, col="red", lty=2)
    dev.off()
  }
  cat("\n<center>\n")
  cat(paste0("![](",EigenResName,")"),"\n")
  cat("\n</center>\n")
  cat("\n<center>\n")
  cat(paste0("![](",QQPlotName,")"),"\n")
  cat("\n</center>\n")

  EigenRes$Espace = rep(i,dim(EigenRes)[1])
  tophit = rbind(tophit, EigenRes[order(EigenRes$P),][c(1:10),c("Espace","CHR","SNP","BP","P","Praw")])
}

tophit[,c(5,6)] = format(tophit[,c(5,6)],digits = 4)
# When kable() is wrapped insiade other expression such as for loop, print(kable()) must be explicit
knitr::kable(tophit,
             caption = "Top hits in each espace", 
             col.names = c("Espace", "CHR", "SNP", "BP", "P", "Praw"),
             row.names = FALSE,
             align = 'c')

```

## 6 Citation
[Chen, G.B. et al, EigenGWAS: finding loci under selection through genome-wide association studies of eigenvectors in structured populations, Heredity, 2016, 117:51-61.](https://www.nature.com/articles/hdy201625)

[Guo-An Qi et al, EigenGWAS: An online visualizing and interactive application for detecting genomic signatures of natural selection, Molecular Ecology Resources, 2021, 00:1-13.](https://onlinelibrary.wiley.com/doi/10.1111/1755-0998.13370)

## 7 Logs
```{r, logs,echo=FALSE}
nn=nrow(read.table(paste0(params$froot, ".fam"), as.is = T))
mm=nrow(read.table(paste0(params$froot, ".bim"), as.is = T))
popTp = ifelse(params$sc==1,"outbred","inbred")
cat(paste0("Input file: "),params$uploadfile)
cat(paste0("Population type: ",popTp))
cat(paste0("Sample size: ",nn))
cat(paste0("Markers: ",mm))
cat(paste0("Marker proportion for quick eigen-decomposition: ",params$proportion*100,"%"))
cat(paste0("Scanned eigen spaces: ",params$espace))
cat(paste0("Report generated at: ",Sys.time()," CHN"))
```
